S4TF_Tutorial_3.ipynb

```diff
@@ -0,0 +1,1772 @@
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "Urigd4SAAtQZ",
+        "colab_type": "text"
+      },
+      "source": [
+        "### Process the data\n",
+        "\n",
+        "Let's explore the format of the dataset before training the model. The following shows there are 60,000 images in the training set, and 10000 images in the test set:"
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "mE76dHtHAumU",
+        "colab_type": "code",
+        "outputId": "a45ba6d3-bbb8-49f2-ceb8-87dffc181d84",
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 119
+        }
+      },
+      "source": [
+        "let (trainImages, trainNumericLabels, numTrainExamples) = readMNIST(imagesFile: \"train-images-idx3-ubyte?raw=true\",\n",
+        "                                                  labelsFile: \"train-labels-idx1-ubyte?raw=true\")\n",
+        "let trainLabels = Tensor<Float>(oneHotAtIndices: trainNumericLabels, depth: 10)\n",
+        "\n",
+        "let (testImages, testNumericLabels, numTestExamples) = readMNIST(imagesFile: \"t10k-images-idx3-ubyte?raw=true\",\n",
+        "                                                labelsFile: \"t10k-labels-idx1-ubyte?raw=true\")\n",
+        "let testLabels = Tensor<Float>(oneHotAtIndices: testNumericLabels, depth: 10)\n",
+        "\n",
+        "\n",
+        "print(\"Number of training examples: \\(numTrainExamples)\")\n",
+        "\n",
+        "print(\"Number of test examples: \\(numTestExamples)\")"
+      ],
+      "execution_count": 7,
+      "outputs": [
+        {
+          "output_type": "stream",
+          "text": [
+            "Reading data.\n",
+            "Constructing data tensors.\n",
+            "Reading data.\n",
+            "Constructing data tensors.\n",
+            "Number of training examples: 60000\n",
+            "Number of test examples: 10000\n"
+          ],
+          "name": "stdout"
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "whCyfZgfCGVl",
+        "colab_type": "text"
+      },
+      "source": [
+        "### Explore the processed data\n",
+        "\n",
+        "Let's plot an image to see what it looks like."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "pXyFCHQTCJMY",
+        "colab_type": "code",
+        "outputId": "fbe55efd-72e1-4813-b505-3203de6dfdd9",
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 281
+        }
+      },
+      "source": [
+        "let firstTrainImage = trainImages[0]\n",
+        "let firstTrainLabel: Int = Int((trainNumericLabels[0].scalars)[0])\n",
+        "let numpyArray = firstTrainImage.makeNumpyArray().reshape(28, 28)\n",
+        "\n",
+        "//print(\"firstTrainImage type:\", type(of: firstTrainImage))\n",
```

Comment:
The comments here (and in examples that follow) are probably leftover debugging comments, so those could all be removed before final commit of this notebook to remove that debugging remnant.

Reply:
I left them intentionally, so that people who want to understand better can comment them off and see what are the shapes and types being used for processing.

Reply:
@BradLarson Should I remove them or let them be?

Reply:
Honestly, I'd remove these to focus on the core concepts being shown. If you wanted to call out the ability to inspect type, shape, etc., it might be worthwhile to have that in a separate cell with accompanying text. In this notebook, I think you're fine without the commented code.